azdev/operations/setup.py

```diff
@@ -245,14 +250,169 @@ def add_ext_repo(path):
         raise CLIError('Installation aborted.')
 
 
-def setup(cli_path=None, ext_repo_path=None, ext=None, deps=None):
+def _validate_input(cli_path, ext_repo_path, set_env, copy, use_global, ext):
+    if copy and use_global:
+        raise CLIError("Copy and use global are mutally exlcusive.")
+    if cli_path == "pypi" and any([use_global, copy, set_env]):
+        raise CLIError("pypi for cli path is mutally exlcusive with global copy and set env")
+    if not cli_path and any([use_global, copy, set_env]):
+        raise CLIError("if global, copy, or set env are set then both an extensions repo "
+                       " and a cli repo must be specified")
+    if not ext_repo_path and ext:
+        raise CLIError("Extesions provided to be installed but no extensions path was given")
+
+
+def _check_paths(cli_path, ext_repo_path):
+    if not os.path.isdir(cli_path):
+        raise CLIError("The cli path is not a valid directory, please check the path")
+    if ext_repo_path and not os.path.isdir(ext_repo_path):
+        raise CLIError("The cli extensions path is not a valid directory, please check the path")
+
+
+def _check_shell():
+    if 'SHELL' in os.environ and const.IS_WINDOWS and 'bash.exe' in os.environ['SHELL']:
+        heading("WARNING: You are running bash in Windows, the setup may not work correctly and "
+                "command may have unexpected behavior")
+        from knack.prompting import prompt_y_n
+        if not prompt_y_n('Would you like to continue with the install?'):
+            sys.exit(0)
+
+
+def _check_env(set_env):
+    if not set_env:
+        if not get_env_path():
+            raise CLIError('You are not running in a virtual enviroment and have not chosen to set one up.')
+    elif 'VIRTUAL_ENV' in os.environ:
+        raise CLIError("You are already running in a virtual enviroment, yet you want to set a new one up")
 
-    require_virtual_env()
 
-    start = time.time()
+def _check_pyenv():
+    if 'PYENV_VIRTUAL_ENV' in os.environ and not const.IS_WINDOWS:
```

Comment:
May want to print warning if pyenv on windows is detected, since we don't support pyenv-win.

Reply:
Good point! Will do.